lib/src/rules/avoid_returning_this.dart

```diff
@@ -17,6 +17,12 @@ const _details = r'''
 Returning `this` from a method is redundant; Dart has a cascade operator which
 allows method chaining universally.
 
+Returning `this` is allowed for:
+
+- operators
+- methods with a return type different of the current class
```

Comment:
Why is this not linted?

I'm imagining someone doing, for instance,

```
class X implements Stream {
  ...
  Stream asStream() => this;
}
```

which IMO should be linted. `X` should simply be used as `useStream(new X())` rather than `useStream(new X().asStream())`.

I suppose maybe a case where I'd want this to be exceptional is where "return this" is a downcast, ie,

```
class X implements Stream<Object> {
  ...
  Stream<int> asIntStream() => this;
}
```

because the method is doing something -- its casting `this` in a way that isn't safe.

That said, even that method is not necessary, as you can do `useIntStream(new X())` because the downcast occurs in either case. At best you could make the argument that the method serves as a reassurance that its safe to perform such an upcast at least sometimes, whereas many upcasts are likely to be accidents.

Reply:
I haven't change anything regarding this case. I've only add a line to document this behaviour ;)

Feel free to open a new issue to discuss this point. I tend to agree with you for inherited type.